frontend/src/store/controller/connection/backend.ts

```diff
@@ -0,0 +1,31 @@
+import { CallEffect, delay, put, PutEffect, select, SelectEffect } from 'redux-saga/effects';
```

Comment:
If the saga's itself are ok, maybe I was thinking I should move this to `src/store/connection/saga.ts` instead?

Reply:
I will try to prototype a somewhat different approach to implementing the sagas so that it matches the behavior in the backend more closely, but yes eventually we should be moving several things from src/store/controller into src/store/connection.

Reply:
The solution I ended up with seems to look roughly similar to your approach, except I make the race between heartbeat update and heartbeat timeout more explicit, using redux-saga's [race](https://redux-saga.js.org/docs/advanced/RacingEffects), and I make the heartbeat update more directly linked to receiving a State Segment. I suspect your approach would also work without your `while (true)` loops in `store/controller/connection/backend.ts`, which makes the intent of the implementation harder for me to follow.

In 12dec45 I also added commented-out code in the Python backend so that it sends states to the frontend for 10 s, then stops sending states for 10 s, then sends states again for 10 s, etc.; this makes it possible to test the heartbeat timeout detection/recovery behavior in the frontend.

Reply:
Oh awesome! so we only use `race` if the frontend is frozen or there is a delay of sorts and we dispatch lost backend action on loss of WebSocket. works as expected on my end!